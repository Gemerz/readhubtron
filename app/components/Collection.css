.listCotent {
    flex: 1 0 280px;
}

.toolbar {
    position: absolute;
    width: calc(100% - 30px);
    height: 60px;
    top: 0;
    left: 0;
    text-align: right;
    padding: 0 15px;
    display: flex;
    flex-direction: row;
    flex-wrap: nowrap;
    justify-content: flex-end;
    align-items: center;
}

.toolbar i {
    font-size: 18px;
}

.refreshBadge span {
    width: 20px;
    height: 20px;
    font-size: 9px;
    top: -8px;
    right: -15px;
}

.refreshBadge img {
    width: 30px;
    height: 30px;
}

.hideNumber span {
    display: none;
}

.list {
    background: #f9f9f9;
    width: 100%;
    height: 100%;
    color: #000;
    padding: 50px 0 0 0;
}

.collection {
    display: block;
    width: 100%;
    height: calc(100% - 50px);
    font-size: 12px;
}

.collection p {
    font-size: 12px;
}

.collectionItem {
    margin: 0;
    padding: 12px 15px;
    border-bottom: 1px solid #ddd;
    cursor: pointer;
}

.collectionTitle {
    font-size: 15px;
}

.collectionSummary {
    font-size: 13px;
    color: #9a9a9a;
    padding: 10px 0;
}

.fromNow {
    color: #9a9a9a;
    text-align: right;
    padding: 5px 20px 10px 20px;
    display: flex;
    justify-content:space-between;
    align-items: center;

}

.loadmore {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    height: 40px;
}
.loading{
  position: absolute;
  width: 100%;
  height: calc(100% - 60px);
  background: #f9f9f9;
  z-index: 1000;
  top:60px;
  left:0;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}
.loading.hide{
  display: none;
}