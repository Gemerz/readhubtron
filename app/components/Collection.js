import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Scrollbars } from 'react-custom-scrollbars';
import Divider from 'material-ui/Divider';
import List, { ListItem, ListItemText } from 'material-ui/List';
import Moment from 'moment';
import { CircularProgress } from 'material-ui/Progress';
import { IconButton, Badge } from 'material-ui';

import 'material-ui/Form/FormHelperText';
import 'material-ui/colors/lightBlue';
import styles from './Collection.css';

Moment.locale('zh-CN');

export default class Collection extends Component {

  state = {
    hasScroll: false,
  }

  get getClientHeight() {
    return this.scrollBars.getClientHeight();
  }
  get getScrollTop() {
    return this.scrollBars.getScrollTop();
  }
  get getScrollHeight() {
    return this.scrollBars.getScrollHeight();
  }
  handlerRefresh() {
    this.props.initList();
    this.props.fetchLatestCollection(this.props.lastCursor);
    this.scrollBars.scrollToTop();
  }
  renderNewCollection() {
    const { collection, setting } = this.props;
    return collection.map((item) => {
      const clickUrl = () => {
        let topicUrl = '';
        if (item.newsArray) { topicUrl = item.newsArray[0].mobileUrl ? item.newsArray[0].mobileUrl : ''; }
        const singleUrl = setting.moblieFirst ? item.mobileUrl : item.url;
        switch (this.props.category) {
          case 'topic':
            return topicUrl;
          case 'news':
            return singleUrl;
          case 'tech':
            return singleUrl;
          default:
            return topicUrl;
        }
      };
      const setCurrentUrlAction = () => {
        switch (this.props.category) {
          case 'topic':
            return this.props.setCurrentUrl(clickUrl())
              && this.props.setTopicTotalUrls(item.newsArray);
          case 'news':
            return this.props.setCurrentUrl(clickUrl());
          case 'tech':
            return this.props.setCurrentUrl(clickUrl());
          default:
            return this.props.setCurrentUrl(clickUrl());
        }
      };
      return (
        <div key={item.id}>
          <ListItem
            button
            className={`${styles.collectionItem} collection-item`}
            data-tid="collectionItem"
            onClick={() => { setCurrentUrlAction(); }}
          >

            <ListItemText primary={item.title} secondary={setting.simpleMode ? '' : item.summary} />
          </ListItem>
          <div className={`${styles.fromNow}`} data-tid="fromNow">
            <span>{item.siteName}</span>
            <span> {Moment(item.publishDate).fromNow()}</span>
          </div>
          <Divider light />
        </div>
      );
    });
  }

  render() {
    return (
      <div className={`${styles.listCotent}`} data-tid="listCotent">

        <div className={` ${styles.toolbar}`} data-tid="toolbar">
          <IconButton
            color="default"
            aria-label="refresh"
            onClick={this.handlerRefresh.bind(this)}
          >
            <Badge
              className={`${this.props.count > 0 ? '' : styles.hideNumber} ${styles.refreshBadge}`}
              data-tid="refreshBadge"
              badgeContent={
                this.props.count}
              color="primary"
            >
              <i className="fa fa-refresh" aria-hidden="true" />
            </Badge>
          </IconButton>
        </div>

        <div className={styles.list} data-tid="list">
          <div className={styles.collection} data-tid="collection">
            <Scrollbars
              ref={(c) => { this.scrollBars = c; }}
              autoHide
              style={{ width: 280 }}
              onScroll={() => {
                if (parseFloat(this.getScrollTop) >=
                  parseFloat(this.getScrollHeight - this.getClientHeight)) {
                  // const hasScroll = false;
                  if (!this.state.hasScroll) {
                    this.props.loadMoreList();
                    this.setState({ hasScroll: true });
                    setTimeout(() => {
                      this.setState({ hasScroll: false });
                    }, 500);
                  }
                }
              }}
            >
              <List className="collection-list">
                {this.renderNewCollection()}
              </List>
              <div className={`${this.props.moreLoading ? '' : 'hide'} ${styles.loadmore}`} data-tid="loadmore">
                <CircularProgress size={20} />
              </div>
            </Scrollbars>

          </div>
        </div>
      </div>
    );
  }
}
Collection.propTypes = {
  initList: PropTypes.func.isRequired,
  fetchLatestCollection: PropTypes.func.isRequired,
  setCurrentUrl: PropTypes.func.isRequired,
  lastCursor: PropTypes.number.isRequired,
  category: PropTypes.string.isRequired,
  setTopicTotalUrls: PropTypes.func,
  setting: PropTypes.shape({
    moblieFirst: PropTypes.bool,
    simpleMode: PropTypes.bool
  }).isRequired,
  count: PropTypes.number.isRequired,
  loadMoreList: PropTypes.func.isRequired,
  moreLoading: PropTypes.bool.isRequired,
  collection: PropTypes.oneOfType([PropTypes.array, PropTypes.object]).isRequired
};

Collection.defaultProps = {
  setTopicTotalUrls: null
};

