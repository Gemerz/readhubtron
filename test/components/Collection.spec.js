import { spy } from 'sinon';
import React from 'react';
import { shallow } from 'enzyme';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import Moment from 'moment';
import Collection from '../../app/components/Collection';
import { Scrollbars } from 'react-custom-scrollbars';

function setup() {
  const actions = {
    initList: spy(),
    fetchLatestCollection: spy(),
    setCurrentUrl: spy(),
    loadMoreList: spy()
  };

  const component = shallow(
    <Collection
      collection={[{ id: '1' }, { id: '2' }]}
      lastCursor={2011}
      count={3}
      moreLoading={false}
      category={'topic'}
      setting={{
        moblieFirst: true,
        simpleMode: true
      }}
      {...actions} />);
  return {
    component,
    actions,
    list: component.find('.collection-item'),
    button: component.find('button')
  };
}

describe('Collection component', () => {
  it('should render component', () => {
    const { component } = setup();
    expect(component).toMatchSnapshot();
  });

  it('render list ', async () => {
    const { list } = setup();
    expect(list.length).toBe(2);

  });
  it(' scroll load more list ', async () => {
    const { list, component, actions } = setup();

    window.dispatchEvent(new window.UIEvent('scroll', { detail: 1000 }));
    component.setState({ hasScroll: true });
    expect(component.state().hasScroll).toEqual(true);
    component.instance().props.loadMoreList();
    expect(actions.loadMoreList.called).toBe(true);
    component.setProps({ collection: [{ id: '1' }, { id: '2' }, { id: '3' }] });
    component.update();
    expect(component.find('.collection-item').length).toBe(3);
  });
  it('click refresh btn', async () => {
    const { button, component, actions } = setup();
    button.at(0).simulate('click');
    component.instance().props.initList();
    component.instance().props.fetchLatestCollection();
    expect(actions.initList.called).toBe(true);
    expect(actions.fetchLatestCollection.called).toBe(true);
  })
});
