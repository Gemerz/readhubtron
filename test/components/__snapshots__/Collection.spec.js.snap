// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection component should render component 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="listCotent"
    data-tid="listCotent"
>
    <div
        className=" toolbar"
        data-tid="toolbar"
    >
        <withStyles(IconButton)
            aria-label="refresh"
            color="default"
            onClick={[Function]}
        >
            <withStyles(Badge)
                badgeContent={3}
                className=" refreshBadge"
                color="primary"
                data-tid="refreshBadge"
            >
                <i
                    aria-hidden="true"
                    className="fa fa-refresh"
                />
            </withStyles(Badge)>
        </withStyles(IconButton)>
    </div>
    <div
        className="list"
        data-tid="list"
    >
        <div
            className="collection"
            data-tid="collection"
        >
            <Scrollbars
                autoHeight={false}
                autoHeightMax={200}
                autoHeightMin={0}
                autoHide={true}
                autoHideDuration={200}
                autoHideTimeout={1000}
                hideTracksWhenNotNeeded={false}
                onScroll={[Function]}
                renderThumbHorizontal={[Function]}
                renderThumbVertical={[Function]}
                renderTrackHorizontal={[Function]}
                renderTrackVertical={[Function]}
                renderView={[Function]}
                style={
                    Object {
                        "width": 280,
                      }
                }
                tagName="div"
                thumbMinSize={30}
                universal={false}
            >
                <withStyles(List)
                    className="collection-list"
                >
                    <div>
                        <withStyles(ListItem)
                            button={true}
                            className="collectionItem collection-item"
                            data-tid="collectionItem"
                            onClick={[Function]}
                        >
                            <withStyles(ListItemText)
                                primary={undefined}
                                secondary=""
                            />
                        </withStyles(ListItem)>
                        <div
                            className="fromNow"
                            data-tid="fromNow"
                        >
                            <span />
                            <span>
                                 
                                几秒前
                            </span>
                        </div>
                        <withStyles(Divider)
                            light={true}
                        />
                    </div>
                    <div>
                        <withStyles(ListItem)
                            button={true}
                            className="collectionItem collection-item"
                            data-tid="collectionItem"
                            onClick={[Function]}
                        >
                            <withStyles(ListItemText)
                                primary={undefined}
                                secondary=""
                            />
                        </withStyles(ListItem)>
                        <div
                            className="fromNow"
                            data-tid="fromNow"
                        >
                            <span />
                            <span>
                                 
                                几秒前
                            </span>
                        </div>
                        <withStyles(Divider)
                            light={true}
                        />
                    </div>
                </withStyles(List)>
                <div
                    className="hide loadmore"
                    data-tid="loadmore"
                >
                    <withStyles(CircularProgress)
                        size={20}
                    />
                </div>
            </Scrollbars>
        </div>
    </div>
</div>,
  "nodes": Array [
    <div
      className="listCotent"
      data-tid="listCotent"
>
      <div
            className=" toolbar"
            data-tid="toolbar"
      >
            <withStyles(IconButton)
                  aria-label="refresh"
                  color="default"
                  onClick={[Function]}
            >
                  <withStyles(Badge)
                        badgeContent={3}
                        className=" refreshBadge"
                        color="primary"
                        data-tid="refreshBadge"
                  >
                        <i
                              aria-hidden="true"
                              className="fa fa-refresh"
                        />
                  </withStyles(Badge)>
            </withStyles(IconButton)>
      </div>
      <div
            className="list"
            data-tid="list"
      >
            <div
                  className="collection"
                  data-tid="collection"
            >
                  <Scrollbars
                        autoHeight={false}
                        autoHeightMax={200}
                        autoHeightMin={0}
                        autoHide={true}
                        autoHideDuration={200}
                        autoHideTimeout={1000}
                        hideTracksWhenNotNeeded={false}
                        onScroll={[Function]}
                        renderThumbHorizontal={[Function]}
                        renderThumbVertical={[Function]}
                        renderTrackHorizontal={[Function]}
                        renderTrackVertical={[Function]}
                        renderView={[Function]}
                        style={
                              Object {
                                    "width": 280,
                                  }
                        }
                        tagName="div"
                        thumbMinSize={30}
                        universal={false}
                  >
                        <withStyles(List)
                              className="collection-list"
                        >
                              <div>
                                    <withStyles(ListItem)
                                          button={true}
                                          className="collectionItem collection-item"
                                          data-tid="collectionItem"
                                          onClick={[Function]}
                                    >
                                          <withStyles(ListItemText)
                                                primary={undefined}
                                                secondary=""
                                          />
                                    </withStyles(ListItem)>
                                    <div
                                          className="fromNow"
                                          data-tid="fromNow"
                                    >
                                          <span />
                                          <span>
                                                 
                                                几秒前
                                          </span>
                                    </div>
                                    <withStyles(Divider)
                                          light={true}
                                    />
                              </div>
                              <div>
                                    <withStyles(ListItem)
                                          button={true}
                                          className="collectionItem collection-item"
                                          data-tid="collectionItem"
                                          onClick={[Function]}
                                    >
                                          <withStyles(ListItemText)
                                                primary={undefined}
                                                secondary=""
                                          />
                                    </withStyles(ListItem)>
                                    <div
                                          className="fromNow"
                                          data-tid="fromNow"
                                    >
                                          <span />
                                          <span>
                                                 
                                                几秒前
                                          </span>
                                    </div>
                                    <withStyles(Divider)
                                          light={true}
                                    />
                              </div>
                        </withStyles(List)>
                        <div
                              className="hide loadmore"
                              data-tid="loadmore"
                        >
                              <withStyles(CircularProgress)
                                    size={20}
                              />
                        </div>
                  </Scrollbars>
            </div>
      </div>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Collection
        category="topic"
        collection={
                Array [
                        Object {
                          "id": "1",
                        },
                        Object {
                          "id": "2",
                        },
                      ]
        }
        count={3}
        fetchLatestCollection={[Function]}
        initList={[Function]}
        lastCursor={2011}
        loadMoreList={[Function]}
        moreLoading={false}
        setCurrentUrl={[Function]}
        setTopicTotalUrls={null}
        setting={
                Object {
                        "moblieFirst": true,
                        "simpleMode": true,
                      }
        }
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Collection {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "category": "topic",
          "collection": Array [
            Object {
              "id": "1",
            },
            Object {
              "id": "2",
            },
          ],
          "count": 3,
          "fetchLatestCollection": [Function],
          "initList": [Function],
          "lastCursor": 2011,
          "loadMoreList": [Function],
          "moreLoading": false,
          "setCurrentUrl": [Function],
          "setTopicTotalUrls": null,
          "setting": Object {
            "moblieFirst": true,
            "simpleMode": true,
          },
        },
        "refs": Object {},
        "state": Object {
          "hasScroll": false,
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="listCotent"
          data-tid="listCotent"
>
          <div
                    className=" toolbar"
                    data-tid="toolbar"
          >
                    <withStyles(IconButton)
                              aria-label="refresh"
                              color="default"
                              onClick={[Function]}
                    >
                              <withStyles(Badge)
                                        badgeContent={3}
                                        className=" refreshBadge"
                                        color="primary"
                                        data-tid="refreshBadge"
                              >
                                        <i
                                                  aria-hidden="true"
                                                  className="fa fa-refresh"
                                        />
                              </withStyles(Badge)>
                    </withStyles(IconButton)>
          </div>
          <div
                    className="list"
                    data-tid="list"
          >
                    <div
                              className="collection"
                              data-tid="collection"
                    >
                              <Scrollbars
                                        autoHeight={false}
                                        autoHeightMax={200}
                                        autoHeightMin={0}
                                        autoHide={true}
                                        autoHideDuration={200}
                                        autoHideTimeout={1000}
                                        hideTracksWhenNotNeeded={false}
                                        onScroll={[Function]}
                                        renderThumbHorizontal={[Function]}
                                        renderThumbVertical={[Function]}
                                        renderTrackHorizontal={[Function]}
                                        renderTrackVertical={[Function]}
                                        renderView={[Function]}
                                        style={
                                                  Object {
                                                            "width": 280,
                                                          }
                                        }
                                        tagName="div"
                                        thumbMinSize={30}
                                        universal={false}
                              >
                                        <withStyles(List)
                                                  className="collection-list"
                                        >
                                                  <div>
                                                            <withStyles(ListItem)
                                                                      button={true}
                                                                      className="collectionItem collection-item"
                                                                      data-tid="collectionItem"
                                                                      onClick={[Function]}
                                                            >
                                                                      <withStyles(ListItemText)
                                                                                primary={undefined}
                                                                                secondary=""
                                                                      />
                                                            </withStyles(ListItem)>
                                                            <div
                                                                      className="fromNow"
                                                                      data-tid="fromNow"
                                                            >
                                                                      <span />
                                                                      <span>
                                                                                 
                                                                                几秒前
                                                                      </span>
                                                            </div>
                                                            <withStyles(Divider)
                                                                      light={true}
                                                            />
                                                  </div>
                                                  <div>
                                                            <withStyles(ListItem)
                                                                      button={true}
                                                                      className="collectionItem collection-item"
                                                                      data-tid="collectionItem"
                                                                      onClick={[Function]}
                                                            >
                                                                      <withStyles(ListItemText)
                                                                                primary={undefined}
                                                                                secondary=""
                                                                      />
                                                            </withStyles(ListItem)>
                                                            <div
                                                                      className="fromNow"
                                                                      data-tid="fromNow"
                                                            >
                                                                      <span />
                                                                      <span>
                                                                                 
                                                                                几秒前
                                                                      </span>
                                                            </div>
                                                            <withStyles(Divider)
                                                                      light={true}
                                                            />
                                                  </div>
                                        </withStyles(List)>
                                        <div
                                                  className="hide loadmore"
                                                  data-tid="loadmore"
                                        >
                                                  <withStyles(CircularProgress)
                                                            size={20}
                                                  />
                                        </div>
                              </Scrollbars>
                    </div>
          </div>
</div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="listCotent"
          data-tid="listCotent"
>
          <div
                    className=" toolbar"
                    data-tid="toolbar"
          >
                    <withStyles(IconButton)
                              aria-label="refresh"
                              color="default"
                              onClick={[Function]}
                    >
                              <withStyles(Badge)
                                        badgeContent={3}
                                        className=" refreshBadge"
                                        color="primary"
                                        data-tid="refreshBadge"
                              >
                                        <i
                                                  aria-hidden="true"
                                                  className="fa fa-refresh"
                                        />
                              </withStyles(Badge)>
                    </withStyles(IconButton)>
          </div>
          <div
                    className="list"
                    data-tid="list"
          >
                    <div
                              className="collection"
                              data-tid="collection"
                    >
                              <Scrollbars
                                        autoHeight={false}
                                        autoHeightMax={200}
                                        autoHeightMin={0}
                                        autoHide={true}
                                        autoHideDuration={200}
                                        autoHideTimeout={1000}
                                        hideTracksWhenNotNeeded={false}
                                        onScroll={[Function]}
                                        renderThumbHorizontal={[Function]}
                                        renderThumbVertical={[Function]}
                                        renderTrackHorizontal={[Function]}
                                        renderTrackVertical={[Function]}
                                        renderView={[Function]}
                                        style={
                                                  Object {
                                                            "width": 280,
                                                          }
                                        }
                                        tagName="div"
                                        thumbMinSize={30}
                                        universal={false}
                              >
                                        <withStyles(List)
                                                  className="collection-list"
                                        >
                                                  <div>
                                                            <withStyles(ListItem)
                                                                      button={true}
                                                                      className="collectionItem collection-item"
                                                                      data-tid="collectionItem"
                                                                      onClick={[Function]}
                                                            >
                                                                      <withStyles(ListItemText)
                                                                                primary={undefined}
                                                                                secondary=""
                                                                      />
                                                            </withStyles(ListItem)>
                                                            <div
                                                                      className="fromNow"
                                                                      data-tid="fromNow"
                                                            >
                                                                      <span />
                                                                      <span>
                                                                                 
                                                                                几秒前
                                                                      </span>
                                                            </div>
                                                            <withStyles(Divider)
                                                                      light={true}
                                                            />
                                                  </div>
                                                  <div>
                                                            <withStyles(ListItem)
                                                                      button={true}
                                                                      className="collectionItem collection-item"
                                                                      data-tid="collectionItem"
                                                                      onClick={[Function]}
                                                            >
                                                                      <withStyles(ListItemText)
                                                                                primary={undefined}
                                                                                secondary=""
                                                                      />
                                                            </withStyles(ListItem)>
                                                            <div
                                                                      className="fromNow"
                                                                      data-tid="fromNow"
                                                            >
                                                                      <span />
                                                                      <span>
                                                                                 
                                                                                几秒前
                                                                      </span>
                                                            </div>
                                                            <withStyles(Divider)
                                                                      light={true}
                                                            />
                                                  </div>
                                        </withStyles(List)>
                                        <div
                                                  className="hide loadmore"
                                                  data-tid="loadmore"
                                        >
                                                  <withStyles(CircularProgress)
                                                            size={20}
                                                  />
                                        </div>
                              </Scrollbars>
                    </div>
          </div>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Collection
    category="topic"
    collection={
        Array [
            Object {
              "id": "1",
            },
            Object {
              "id": "2",
            },
          ]
    }
    count={3}
    fetchLatestCollection={[Function]}
    initList={[Function]}
    lastCursor={2011}
    loadMoreList={[Function]}
    moreLoading={false}
    setCurrentUrl={[Function]}
    setTopicTotalUrls={null}
    setting={
        Object {
            "moblieFirst": true,
            "simpleMode": true,
          }
    }
/>,
}
`;
